frontend/src/pages/matching-service/FindingMatch.js

```diff
@@ -0,0 +1,188 @@
+  // const [animationKey, setAnimationKey] = useState(0); // Key to reset the animation effect
+  // let typingInterval;
+  // const fullText = "  Matching in progress.... ";
+  const navigate = useNavigate();
+  const location = useLocation(); // Use useLocation to retrieve state
+  const { topic, difficultyLevel, email, token, username } = location.state || {}; // Destructure updatedFormData from state
+  const [isAnyDifficulty, setIsAnyDifficulty] = useState(false);
+
+  // check for backtrack, navigate back to criteria selection if user confirms action,
+  // otherwise stay on page
+  window.onpopstate = (event) => {
+    window.alert("Going back to criteria selection...");
+    socket.emit("cancel_matching", { topic, difficultyLevel, email, token, username, isAny: isAnyDifficulty });
+    navigate("/matching/select");
+    // try again later
+    // var confirmation = window.confirm("You are exiting the matching queue, continue?");
+    // if (confirmation) {
+    //   socket.emit("cancel_matching", { topic, difficultyLevel, email, token, username, isAny: isAnyDifficulty });
+    //   location.state = undefined;
+    //   navigate("/matching/select");
+    // } else {
+    //   event.stopImmediatePropagation();
+    //   event.preventDefault();
+    //   socket.emit("cancel_matching", { topic, difficultyLevel, email, token, username, isAny: isAnyDifficulty });
+    //   navigate("/matching/select");
+    // }
+  }
+
+  // check for refresh, allow user to stay on page regardless if refresh is cancelled or confirmed
+  window.onbeforeunload = (event) => {
+    location.state = undefined;
+    socket.emit("cancel_matching", { topic, difficultyLevel, email, token, username, isAny: isAnyDifficulty });
+  }
+
+  // if refresh and the page reloads, send user back to queue
+  window.onload = (event) => {
+    socket.emit("join_matching_queue", { topic, difficultyLevel, email, token, username, isAny: isAnyDifficulty });
+  }
+
+  // detect changes for isAnyDifficulty (used for cancelling queue)
+  useEffect(() => {
+    setIsAnyDifficulty((prevState) => !prevState);
+  }, []);
+
+  useEffect(() => {
+    console.log("changing isAnyDifficulty", isAnyDifficulty);
+  }, [isAnyDifficulty]);
+
+  // Timer effect
+  useEffect(() => {
+    const timer = setInterval(() => {
+      setTimeLeft((prevTime) => {
+        if (prevTime <= 1) {
+          clearInterval(timer);  // Stop the timer when it reaches zero
+          if (matchStatus !== "Matching cancelled") {
+            setMatchStatus("No match found"); // Set status when timer reaches 0
+          }
+          console.log("Matching failed due to timeout");
+          return 0;
+        }
+        return prevTime - 1;
+      });
+    }, 1000); // Update every second
+
+    return () => clearInterval(timer); // Cleanup timer on component unmount
+  }, [matchStatus]); // Restart the timer whenever matchStatus or animationKey changes
+
+  // // Loading animation 
```

Comment:
maybe this part can remove if not using

Reply:
keeping for now in case we want to make the loading animation work properly